core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -120,29 +125,42 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
     }
   }
 
-  def updateCache(updateMetadataRequest: UpdateMetadataRequest,
+  def updateCache(correlationId: Int, updateMetadataRequest: UpdateMetadataRequest,
                   brokerId: Int,
                   stateChangeLogger: StateChangeLogger) {
     inWriteLock(partitionMetadataLock) {
-      aliveBrokers = updateMetadataRequest.aliveBrokers.map(b => (b.id, b)).toMap
-      updateMetadataRequest.partitionStateInfos.foreach { case(tp, info) =>
-        if (info.leaderIsrAndControllerEpoch.leaderAndIsr.leader == LeaderAndIsr.LeaderDuringDelete) {
+      aliveBrokers = updateMetadataRequest.liveBrokers.asScala.map { broker =>
+        val endPoints = broker.endPoints.asScala.map { case (protocol, ep) =>
+          (protocol,EndPoint(ep.host, ep.port, protocol))
+        }.toMap
+        (broker.id, Broker(broker.id, endPoints))
+      }.toMap
+
+      updateMetadataRequest.partitionStates.asScala.foreach { case(tp, info) =>
+        if (info.leader == LeaderAndIsr.LeaderDuringDelete) {
           removePartitionInfo(tp.topic, tp.partition)
           stateChangeLogger.trace(("Broker %d deleted partition %s from metadata cache in response to UpdateMetadata request " +
             "sent by controller %d epoch %d with correlation id %d")
             .format(brokerId, tp, updateMetadataRequest.controllerId,
-            updateMetadataRequest.controllerEpoch, updateMetadataRequest.correlationId))
+              updateMetadataRequest.controllerEpoch, correlationId))
         } else {
-          addOrUpdatePartitionInfo(tp.topic, tp.partition, info)
+          val partitionInfo = partitionStateToPartitionStateInfo(info)
+          addOrUpdatePartitionInfo(tp.topic, tp.partition, partitionInfo)
           stateChangeLogger.trace(("Broker %d cached leader info %s for partition %s in response to UpdateMetadata request " +
             "sent by controller %d epoch %d with correlation id %d")
             .format(brokerId, info, tp, updateMetadataRequest.controllerId,
-            updateMetadataRequest.controllerEpoch, updateMetadataRequest.correlationId))
+              updateMetadataRequest.controllerEpoch, correlationId))
         }
       }
     }
   }
 
+  private def partitionStateToPartitionStateInfo(partitionState: PartitionState): PartitionStateInfo = {
+    val leaderAndIsr = LeaderAndIsr(partitionState.leader, partitionState.leaderEpoch, partitionState.isr.asScala.map(_.toInt).toList, partitionState.zkVersion)
+    val leaderInfo = LeaderIsrAndControllerEpoch(leaderAndIsr, partitionState.controllerEpoch)
+    PartitionStateInfo(leaderInfo, partitionState.replicas.asScala.map(_.toInt))
```

Comment:
The plan is to migrate the TopicMetadata Request/Response [KAFKA-2073](https://issues.apache.org/jira/browse/KAFKA-2073), and then refactor the metadata cache to optimize usage and try to remove those classes [KAFKA-2969](https://issues.apache.org/jira/browse/KAFKA-2969). This is the pattern I am taking for many of these migrations. Other classes like KafkaApis and ReplicaManager has some code that can be refactored after all (or most) messages are migrated too.